x/ccv/provider/keeper/grpc_query.go

```diff
@@ -333,17 +335,59 @@ func (k Keeper) QueryConsumerValidators(goCtx context.Context, req *types.QueryC
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	if _, found := k.GetConsumerClientId(ctx, consumerId); !found {
-		// chain has to have started; consumer client id is set for a chain during the chain's spawn time
-		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("no started consumer chain: %s", consumerId))
+	// check that the consumer is initialized
+	initParams, err := k.GetConsumerInitializationParameters(ctx, consumerId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, errorsmod.Wrap(types.ErrUnknownConsumerId, consumerId).Error())
 	}
 
-	var validators []*types.QueryConsumerValidatorsValidator
+	// query consumer validator set
 
-	consumerValSet, err := k.GetConsumerValSet(ctx, consumerId)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+	var consumerValSet []types.ConsensusValidator
+	// if the consumer hasn't launched yet, compute the consumer validator set
+	if ctx.BlockTime().Before(*initParams.SpawnTime) {
+		var bondedValidators []stakingtypes.Validator
+
+		powerParams, err := k.GetConsumerPowerShapingParameters(ctx, consumerId)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, types.ErrInvalidConsumerParams.Error())
+		}
+		bondedValidators, err = k.GetLastBondedValidators(ctx)
+		if err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("failed to get last validators: %s", err))
+		}
+
+		if topN := powerParams.GetTop_N(); topN > 0 {
+			// in a Top-N chain, we automatically opt in all validators that belong to the top N
+			// of the active validators
+			activeValidators, err := k.GetLastProviderConsensusActiveValidators(ctx)
+			if err != nil {
+				// something must be broken in the bonded validators, so we have to panic since there is no realistic way to proceed
+				return nil, status.Error(codes.Internal, fmt.Sprintf("failed to get active validators: %s", err))
+			}
+
+			minPower, err := k.ComputeMinPowerInTopN(ctx, activeValidators, topN)
+			if err != nil {
+				// we panic, since the only way to proceed would be to opt in all validators, which is not the intended behavior
+				return nil, status.Error(codes.Internal, fmt.Sprintf("failed to compute min power to opt in for chain %s: %s", consumerId, err))
+			}
+
+			// set the minimal power of validators in the top N in the store
+			k.SetMinimumPowerInTopN(ctx, consumerId, minPower)
+
+			k.OptInTopNValidators(ctx, consumerId, activeValidators, minPower)
+		}
+
+		consumerValSet = k.ComputeNextValidators(ctx, consumerId, bondedValidators)
```

Comment:
It feels that this logic is reused in several places, e.g., here, at the end of every epoch, when initializing a new chain. If that's right, can we have a method to not have to rewrite this code multiple times?

Reply:
I intended to do it, but it seems there is a reason why this logic hasn't been reformated.

@Karolos do you have more context?

Reply:
@sainoe This was a much smaller snippet of code before. We could check if we can modularize it into one function. I think we increased it as part of the inactive-validators work. 
I think we can do this independently of this PR. I put a note under `Potpourri (so we do not forget)` [here](https://github.com/cosmos/interchain-security/issues/2108).